README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# MODISTools <a href='https://github.com/ropensci/MODISTools'><img src='https://raw.githubusercontent.com/ropensci/MODISTools/master/MODISTools-logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![R build
status](https://github.com/ropensci/MODISTools/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/MODISTools/actions)
[![codecov](https://codecov.io/gh/ropensci/MODISTools/branch/master/graph/badge.svg)](https://codecov.io/gh/ropensci/MODISTools)
![Status](https://www.r-pkg.org/badges/version/MODISTools) [![rOpenSci
Peer
Review](https://badges.ropensci.org/246_status.svg)](https://github.com/ropensci/software-review/issues/246)
![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/MODISTools)
<!-- badges: end -->

Programmatic interface to the [‘MODIS Land Products Subsets’ web
services](https://modis.ornl.gov/data/modis_webservice.html). Allows for
easy downloads of [‘MODIS’](http://modis.gsfc.nasa.gov/) time series
directly to your R workspace or your computer. When using the package
please cite the manuscript as referenced below. Keep in mind that the
original manuscript describes versions prior to release 1.0 of the
package. Functions described in this manuscript do not exist in the
current package, please consult [the
documentation](https://docs.ropensci.org/MODISTools/reference/index.html)
to find matching functionality.

## Installation

### stable release

To install the current stable release use a CRAN repository:

``` r
install.packages("MODISTools")
library("MODISTools")
```

### development release

To install the development releases of the package run the following
commands:

``` r
if(!require(devtools)){install.package("devtools")}
devtools::install_github("khufkens/MODISTools")
library("MODISTools")
```

Vignettes are not rendered by default, if you want to include additional
documentation please use:

``` r
if(!require(devtools)){install.package("devtools")}
devtools::install_github("khufkens/MODISTools", build_vignettes = TRUE)
library("MODISTools")
```

## Use

### Downloading MODIS time series

To extract a time series of modis data for a given location and its
direct environment use the mt_subset() function.

<details>
<summary>
detailed parameter description (click to expand)
</summary>
<p>

| Parameter | Description                                                                                                                     |
|-----------|---------------------------------------------------------------------------------------------------------------------------------|
| product   | a MODIS product                                                                                                                 |
| band      | a MODIS product band (if NULL all bands are downloaded)                                                                         |
| lat       | latitude of the site                                                                                                            |
| lon       | longitude of the site                                                                                                           |
| start     | start year of the time series (data start in 1980)                                                                              |
| end       | end year of the time series (current year - 2 years, use force = TRUE to override)                                              |
| internal  | logical, TRUE or FALSE, if true data is imported into R workspace otherwise it is downloaded into the current working directory |
| out_dir   | path where to store the data when not used internally, defaults to tempdir()                                                    |
| km_lr     | force “out of temporal range” downloads (integer)                                                                               |
| km_ab     | suppress the verbose output (integer)                                                                                           |
| site_name | a site identifier                                                                                                               |
| site_id   | a site_id for predefined locations (not required)                                                                               |
| progress  | logical, TRUE or FALSE (show download progress)                                                                                 |

</p>
</details>

``` r
# load the library
library(MODISTools)

# download data
subset <- mt_subset(product = "MOD11A2",
                    lat = 40,
                    lon = -110,
                    band = "LST_Day_1km",
                    start = "2004-01-01",
                    end = "2004-02-01",
                    km_lr = 1,
                    km_ab = 1,
                    site_name = "testsite",
                    internal = TRUE,
                    progress = FALSE)
print(str(subset))
#> 'data.frame':    36 obs. of  21 variables:
#>  $ xllcorner    : chr  "-9370963.05" "-9370963.05" "-9370963.05" "-9370963.05" ...
#>  $ yllcorner    : chr  "4445948.79" "4445948.79" "4445948.79" "4445948.79" ...
#>  $ cellsize     : chr  "926.625433055834" "926.625433055834" "926.625433055834" "926.625433055834" ...
#>  $ nrows        : int  3 3 3 3 3 3 3 3 3 3 ...
#>  $ ncols        : int  3 3 3 3 3 3 3 3 3 3 ...
#>  $ band         : chr  "LST_Day_1km" "LST_Day_1km" "LST_Day_1km" "LST_Day_1km" ...
#>  $ units        : chr  "Kelvin" "Kelvin" "Kelvin" "Kelvin" ...
#>  $ scale        : chr  "0.02" "0.02" "0.02" "0.02" ...
#>  $ latitude     : num  40 40 40 40 40 40 40 40 40 40 ...
#>  $ longitude    : num  -110 -110 -110 -110 -110 -110 -110 -110 -110 -110 ...
#>  $ site         : chr  "testsite" "testsite" "testsite" "testsite" ...
#>  $ product      : chr  "MOD11A2" "MOD11A2" "MOD11A2" "MOD11A2" ...
#>  $ start        : chr  "2004-01-01" "2004-01-01" "2004-01-01" "2004-01-01" ...
#>  $ end          : chr  "2004-02-01" "2004-02-01" "2004-02-01" "2004-02-01" ...
#>  $ complete     : logi  TRUE TRUE TRUE TRUE TRUE TRUE ...
#>  $ modis_date   : chr  "A2004001" "A2004009" "A2004017" "A2004025" ...
#>  $ calendar_date: chr  "2004-01-01" "2004-01-09" "2004-01-17" "2004-01-25" ...
#>  $ tile         : chr  "h09v05" "h09v05" "h09v05" "h09v05" ...
#>  $ proc_date    : chr  "2020168005635" "2020168010833" "2020168012220" "2020168013617" ...
#>  $ pixel        : int  1 1 1 1 2 2 2 2 3 3 ...
#>  $ value        : int  13148 13160 13398 13412 13153 13140 13370 13388 13131 13096 ...
#> NULL
```

The output format is a *tidy* data frame, as shown above. When witten to
a csv with the parameter `internal = FALSE` this will result in a flat
file on disk.

Note that when a a region is defined using km_lr and km_ab multiple
pixels might be returned. These are indexed using the `pixel` column in
the data frame containing the time series data. The remote sensing
values are listed in the `value` column. When no band is specified all
bands of a given product are returned, be mindful of the fact that
different bands might require different multipliers to represent their
true values. To list all available products, bands for particular
products and temporal coverage see function descriptions below.

### Batch downloading MODIS time series

When a large selection of locations is needed you might benefit from
using the batch download function `mt_batch_subset()`, which provides a
wrapper around the `mt_subset()` function in order to speed up large
download batches. This function has a similar syntax to `mt_subset()`
but requires a data frame defining site names (site_name) and locations
(lat / lon) (or a comma delimited file with the same structure) to
specify a list of download locations.

Below an example is provided on how to batch download data for a data
frame of given site names and locations (lat / lon).

``` r
# create data frame with a site_name, lat and lon column
# holding the respective names of sites and their location
df <- data.frame("site_name" = paste("test",1:2))
df$lat <- 40
df$lon <- -110
  
# test batch download
subsets <- mt_batch_subset(df = df,
                     product = "MOD11A2",
                     band = "LST_Day_1km",
                     internal = TRUE,
                     start = "2004-01-01",
                     end = "2004-02-01")

print(str(subsets))
#> 'data.frame':    8 obs. of  21 variables:
#>  $ xllcorner    : chr  "-9370036.35" "-9370036.35" "-9370036.35" "-9370036.35" ...
#>  $ yllcorner    : chr  "4446875.49" "4446875.49" "4446875.49" "4446875.49" ...
#>  $ cellsize     : chr  "926.625433055834" "926.625433055834" "926.625433055834" "926.625433055834" ...
#>  $ nrows        : int  1 1 1 1 1 1 1 1
#>  $ ncols        : int  1 1 1 1 1 1 1 1
#>  $ band         : chr  "LST_Day_1km" "LST_Day_1km" "LST_Day_1km" "LST_Day_1km" ...
#>  $ units        : chr  "Kelvin" "Kelvin" "Kelvin" "Kelvin" ...
#>  $ scale        : chr  "0.02" "0.02" "0.02" "0.02" ...
#>  $ latitude     : num  40 40 40 40 40 40 40 40
#>  $ longitude    : num  -110 -110 -110 -110 -110 -110 -110 -110
#>  $ site         : chr  "test 1" "test 1" "test 1" "test 1" ...
#>  $ product      : chr  "MOD11A2" "MOD11A2" "MOD11A2" "MOD11A2" ...
#>  $ start        : chr  "2004-01-01" "2004-01-01" "2004-01-01" "2004-01-01" ...
#>  $ end          : chr  "2004-02-01" "2004-02-01" "2004-02-01" "2004-02-01" ...
#>  $ complete     : logi  TRUE TRUE TRUE TRUE TRUE TRUE ...
#>  $ modis_date   : chr  "A2004001" "A2004009" "A2004017" "A2004025" ...
#>  $ calendar_date: chr  "2004-01-01" "2004-01-09" "2004-01-17" "2004-01-25" ...
#>  $ tile         : chr  "h09v05" "h09v05" "h09v05" "h09v05" ...
#>  $ proc_date    : chr  "2020168005635" "2020168010833" "2020168012220" "2020168013617" ...
#>  $ pixel        : int  1 1 1 1 1 1 1 1
#>  $ value        : int  13129 13102 13343 13364 13129 13102 13343 13364
#> NULL
```

### Listing products

To list all available products use the mt_products() function.

``` r
products <- mt_products()
head(products)
#>        product
#> 1       Daymet
#> 2 ECO4ESIPTJPL
#> 3      ECO4WUE
#> 4       GEDI03
#> 5     GEDI04_B
#> 6      MCD12Q1
#>                                                                          description
#> 1 Daily Surface Weather Data (Daymet) on a 1-km Grid for North America, Version 4 R1
#> 2               ECOSTRESS Evaporative Stress Index PT-JPL (ESI) Daily L4 Global 70 m
#> 3                          ECOSTRESS Water Use Efficiency (WUE) Daily L4 Global 70 m
#> 4                GEDI Gridded Land Surface Metrics (LSM) L3 1km EASE-Grid, Version 2
#> 5       GEDI Gridded Aboveground Biomass Density (AGBD) L4B 1km EASE-Grid, Version 2
#> 6              MODIS/Terra+Aqua Land Cover Type (LC) Yearly L3 Global 500 m SIN Grid
#>   frequency resolution_meters
#> 1     1 day              1000
#> 2    Varies                70
#> 3    Varies                70
#> 4  One time              1000
#> 5  One time              1000
#> 6    1 year               500
```

### Listing bands

To list all available bands for a given product use the mt_bands()
function.

``` r
bands <- mt_bands(product = "MOD11A2")
head(bands)
#>               band                          description valid_range fill_value
#> 1   Clear_sky_days               Day clear-sky coverage    1 to 255          0
#> 2 Clear_sky_nights             Night clear-sky coverage    1 to 255          0
#> 3    Day_view_angl View zenith angle of day observation    0 to 130        255
#> 4    Day_view_time        Local time of day observation    0 to 240        255
#> 5          Emis_31                   Band 31 emissivity    1 to 255          0
#> 6          Emis_32                   Band 32 emissivity    1 to 255          0
#>    units scale_factor add_offset
#> 1   <NA>         <NA>       <NA>
#> 2   <NA>         <NA>       <NA>
#> 3 degree            1        -65
#> 4    hrs          0.1          0
#> 5   <NA>        0.002       0.49
#> 6   <NA>        0.002       0.49
```

### listing dates

To list all available dates (temporal coverage) for a given product and
location use the mt_dates() function.

``` r
dates <- mt_dates(product = "MOD11A2", lat = 42, lon = -110)
head(dates)
#>   modis_date calendar_date
#> 1   A2000049    2000-02-18
#> 2   A2000057    2000-02-26
#> 3   A2000065    2000-03-05
#> 4   A2000073    2000-03-13
#> 5   A2000081    2000-03-21
#> 6   A2000089    2000-03-29
```

## References

Hufkens (2022). The MODISTools package: an interface to the MODIS Land
Products Subsets Web Services <https://github.com/ropensci/MODISTools>

## Acknowledgements

Original development was supported by the UK Natural Environment
Research Council (NERC; grants NE/K500811/1 and NE/J011193/1), and the
Hans Rausing Scholarship. Refactoring was supported through the Belgian
Science Policy office COBECORE project (BELSPO; grant
BR/175/A3/COBECORE). Logo design elements are taken from the FontAwesome
library according to [these terms](https://fontawesome.com/license),
where the globe element was inverted and intersected. Continued support
for MODISTools is provided by [BlueGreen
Labs](https://bluegreenlabs.org).

<a href='https://bluegreenlabs.org'><img src='https://bluegreenlabs.org/img/logo_text_small.png' width="200"/></a>

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
# MODISTools 1.1.3

* removing {raster} dependencies

# MODISTools 1.1.2

* removing parallel processing in batch tool due to timeout issues
* set error = TRUE on vignette to prevent stopping during render
* correct batch downloading, missing argument in apply()

# MODISTools 1.1.1

* added raster conversion function `mt_to_raster()`
* force integer values on buffer values
* removed keywords

# MODISTools 1.1.0

* ropensci code review
* replaced NULL parameters with missing()
* output a tidy data frame
* include coordinate transform functions to sf objects
* pkgdown documentation generation

# MODISTools 1.0.0

* ropensci refactoring
* version increment follows original package
* push to CRAN

# MODISTools 0.1.0

* first draft


R/mt_subset.R
#' Download MODIS Land Products subsets
#'
#' Download a MODIS Land Products Subset product
#' for a given point location buffered with a given amount of kilometers
#' left-right, top-bottom for a given location (provided as latitude and
#' longitude values).
#'
#' @param product a valid MODIS product name
#' @param band band or bands (as a character vector) to download
#' @param lat latitude in decimal degrees
#' @param lon longitude in decimal degrees
#' @param start start date
#' @param end end date
#' @param km_lr km left-right to sample (rounded to the nearest integer)
#' @param km_ab km above-below to sample (rounded to the nearest integer)
#' @param site_id site id (overides lat / lon)
#' @param network the network for which to generate the site list,
#' when not provided the complete list is provided
#' @param site_name arbitrary site name used in writing data to file
#' (default = sitename)
#' @param out_dir path where to store the data if writing to disk
#' (default = tempdir())
#' @param internal should the data be returned as an internal data structure
#' \code{TRUE} or \code{FALSE} (default = \code{TRUE})
#' @param progress show download progress
#' @return A data frame combining meta-data and actual data values.
#' @seealso \code{\link[MODISTools]{mt_sites}}
#' \code{\link[MODISTools]{mt_dates}} \code{\link[MODISTools]{mt_bands}}
#' \code{\link[MODISTools]{mt_products}}
#' \code{\link[MODISTools]{mt_batch_subset}}
#' @export
#' @examples
#'
#' \donttest{
#' # list all available MODIS Land Products Subsets products
#' # download data
#' subset <- mt_subset(product = "MOD11A2",
#'                         lat = 40,
#'                         lon = -110,
#'                         band = "LST_Day_1km",
#'                         start = "2004-01-01",
#'                         end = "2004-03-31",
#'                         progress = FALSE)
#'  head(subset)
#'}

mt_subset <- function(
  product,
  band,
  lat,
  lon,
  start = "2000-01-01",
  end = format(Sys.time(),"%Y-%m-%d"),
  km_lr = 0,
  km_ab = 0,
  site_id,
  network,
  site_name = "sitename",
  out_dir = tempdir(),
  internal = TRUE,
  progress = TRUE
  ){

  # error trap missing coordinates or site id
  if (missing(site_id) & (missing(lat) | missing(lon)) ){
    stop("please specify coordinates, or a valid site ID...")
  }

  # check if site_id is valid
  if(!missing(site_id)){
    if(missing(network)){

      # load all sites
      sites <- MODISTools::mt_sites()

      # check if the site id is valid
      if (!(site_id %in% sites$siteid)){
        stop("please specify a valid site id...")
      }
    } else {

      # load all sites
      sites <- MODISTools::mt_sites(network = network)

      # check if the site id is valid
      if (!(site_id %in% sites$network_siteid)){
        stop("please specify a valid site id...")
      }
    }
  }

  # load all products
  products <- MODISTools::mt_products()$product

  # error trap product
  if (missing(product) | !(product %in% products) ){
    stop("please specify a product, or check your product name...")
  }

  # load all bands for product
  bands <- mt_bands(product)

  # error trap band
  if (missing(band) | !all(band %in% bands$band) ){
    stop("please specify a band, or check your product band combination ...")
  }

  # switch url in case of siteid
  if (missing(site_id)){
    url <- paste(mt_server(),
                  product,
                  "subset",
                 sep = "/")

    # grab all available dates
    dates <- MODISTools::mt_dates(product = product,
                                  lat = lat,
                                  lon = lon)
  } else {
    if(missing(network)){
      url <- paste(mt_server(),
                    product,
                    site_id,
                    "subset",
                   sep = "/")

      # grab all available dates
      dates <- MODISTools::mt_dates(
        product = product,
        site_id = site_id)
    } else {
      url <- paste(mt_server(),
                   product,
                   network,
                   site_id,
                   "subset",
                   sep = "/")

      # grab all available dates
      dates <- MODISTools::mt_dates(
        product = product,
        network = network,
        site_id = site_id)
    }

    lat <- NULL
    lon <- NULL
  }

  # convert to date object for easier handling
  dates$calendar_date <- as.Date(dates$calendar_date)

  # subset the dates
  dates <- dates[which(dates$calendar_date <= as.Date(end) &
           dates$calendar_date >= as.Date(start)),]

  # check if something remains
  if (nrow(dates) == 0){
    stop("No data points exist for the selected date range...")
  }

  # list breaks, for downloads in chunks
  breaks <- seq(1, nrow(dates), 10)

  # grab site name
  site <- ifelse(missing(site_id), site_name, site_id)

  # create a complete list of the data
  complete_data <- lapply(band, function(band_name){

    # start progress bar chuncks
    if(progress){
      message("Downloading chunks:")
      env <- environment()
      counter <- 0
      pb <- utils::txtProgressBar(
        min = 0,
        max = length(breaks),
        style = 3
        )
    }


    # loop over all 10 value breaks
    subset_data <- lapply(breaks, function(b){

      # grab last date for subset
      if(b == breaks[length(breaks)]){
        end_date <-dates$modis_date[nrow(dates)]
      } else {
        end_date <- dates$modis_date[b+9]
      }

      # construct the query to be served to the server
      query <- list("latitude" = lat,
                    "longitude" = lon,
                    "band" = band_name,
                    "startDate" = dates$modis_date[b],
                    "endDate" = end_date,
                    "kmAboveBelow" = round(km_ab),
                    "kmLeftRight" = round(km_lr))

      # try to download the data
      json_chunk <- httr::RETRY(verb = "GET", 
                                url = url,
                                query = query,
                                httr::write_memory(),
                                terminate_on = c(403, 404))

      # trap errors on download, return a detailed
      # API error statement
      if (httr::http_error(json_chunk)){
        warning(httr::content(json_chunk), call. = FALSE)
        return(NULL)
      }

      # check the content of the response
      if (httr::http_type(json_chunk) != "application/json") {
        warning("API did not return json...", call. = FALSE)
        return(NULL)
      }

      # grab content from cached json chunk
      chunk <- jsonlite::fromJSON(httr::content(json_chunk, "text",
                                                encoding = "UTF-8"),
                                  simplifyVector = TRUE)

      # set progress bar
      if(progress){
        tmp <- get("counter", envir = env)
        assign("counter", tmp + 1 ,envir = env)
        utils::setTxtProgressBar(get("pb", envir = env), tmp + 1)
      }

      # return data
      return(chunk)
    })

    # close progress bar
    if(progress){
      close(pb)
    }

    # split out a header including
    # additional ancillary data
    header <- subset_data[[1]][!(names(subset_data[[1]]) %in%
                                     c("header","subset"))]
    header$site <- site
    header$product <- product
    header$start <- start
    header$end <- end
    header$cellsize <- as.character(header$cellsize)

    # This is a check on the complete nature of the retrieved data
    # the process will not stall on errors occur in the download
    # process but just return NULL, these are trapped here and
    # reported as complete TRUE/FALSE in the header or the returned
    # object. Using this flag one can decide to reprocess.
    header$complete <- !any(unlist(lapply(subset_data, is.null)))


    # reshape the data converting it to a tidy data frame
    # data will be reported row wise
    subset_data <- do.call("rbind",
                           lapply(subset_data,
                                  function(x)x$subset))
    pixels <- do.call("rbind",
                      subset_data$data)
    colnames(pixels) <- seq_len(ncol(pixels))

    # remove old nested list data and substitute with columns
    subset_data <- cbind(subset_data[,!(names(subset_data) %in% "data")],
                              pixels)

    subset_data <- stats::reshape(subset_data,
                           varying = grep("[0-9]",names(subset_data)),
                           direction = "long",
                           timevar = "pixel",
                           v.names = "value")

    # drop the id column
    subset_data <- subset_data[ , !(names(subset_data) %in% "id")]

    # combine header with the data, this repeats
    # some meta-data but makes file handling easier
    subset_data <- data.frame(header, subset_data,
                              stringsAsFactors = FALSE)

    # drop duplicate band column
    subset_data <- subset_data[ , !(names(subset_data) %in% "band.1")]


    # return a nested list with all data
    # to workspace or to file
    if (internal){
      return(subset_data)
    } else {
      # format filename
      filename <- sprintf("%s/%s_%s_%s_%s%s.csv",
                          path.expand(out_dir),
                          header$site,
                          header$product,
                          header$band,
                          header$start,
                          header$end)

      # write file to disk
      utils::write.table(subset_data,
                         filename,
                         quote = FALSE,
                         row.names = FALSE,
                         col.names = TRUE,
                         sep = ",")
    }
  })

 # return a larger tidy dataframe by
 # row binding list elements
 return(do.call("rbind", complete_data))

}
